evm/db/tracked.py

```diff
@@ -0,0 +1,84 @@
+from .backends.base import BaseDB
+
+
+class TrackedDB(BaseDB):
+    """
+    The StateDB would be responsible for collecting all the touched keys.
+
+    reads: the dict of read key-value
+    writes: the dict of written key-value
+    """
+
+    wrapped_db = None
+    _reads = None
+    _writes = None
+
+    def __init__(self, db):
+        self.wrapped_db = db
+        self._reads = {}
+        self._writes = {}
+
+    def get(self, key):
+        """
+        Return the value of specific key and update read dict.
+        """
+        try:
+            current_value = self.wrapped_db.get(key)
+        except KeyError:
+            raise
+        else:
+            self._reads[key] = current_value
+
+        return current_value
+
+    def set(self, key, value):
+        """
+        Set the key-value and update writes dict.
+        """
+        self._writes[key] = value
+        return self.wrapped_db.set(key, value)
+
+    def exists(self, key):
+        """
+        Check if the key exsits.
+        """
+        return self.wrapped_db.exists(key)
+
+    def delete(self, key):
+        """
+        Delete the key and update writes dict.
+        """
+        try:
+            current_value = self.wrapped_db.get(key)
+        except KeyError:
+            # do not need to be added in writes
+            pass
+        else:
+            self._writes[key] = current_value
```

Comment:
Should this instead be set to something like `None` to indicate that it's been deleted?

Reply:
Good, using `None` would be more clear :)
And I think the `delete` function don't *need* to return something?
Can I just modify it to:
```python
     def delete(self, key):
        """
        Delete the key and update writes dict.
        """
        self.wrapped_db.delete(key)
        self._writes[key] = None
```